Wrappers/Python/cil/optimisation/utilities/preconditioner.py

```diff
@@ -0,0 +1,200 @@
+#  Copyright 2024 United Kingdom Research and Innovation
+#  Copyright 2024 The University of Manchester
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+#
+# Authors:
+# - CIL Developers, listed at: https://github.com/TomographicImaging/CIL/blob/master/NOTICE.txt
+
+
+from abc import ABC, abstractmethod
+import numpy as np
+
+
+class Preconditioner(ABC):
+
+    r"""
+    Abstract base class for Preconditioner objects. The `apply` method of this class takes an initialised CIL function as an argument and modifies a provided  `gradient`.
+   
+    
+    Methods
+    -------
+    apply(x)
+        Abstract method to call the preconditioner.
+    """
+
+    def __init__(self):
+        '''Initialises the preconditioner
+        '''
+        pass
+
+    @abstractmethod
+    def apply(self, algorithm, gradient, out):
+        r"""
+        Abstract method to apply the preconditioner.
+
+        Parameters
+        ----------
+        algorithm : Algorithm
+            The algorithm object.
+    
+        gradient : DataContainer
+            The calculated gradient to modify
+        out : DataContainer, 
+            Container to fill with the modified gradient 
+        
+        Returns
+        -------
+        DataContainer
+            The modified gradient
+            
+        """
+        pass
+
+
+class Sensitivity(Preconditioner):
+
+    r"""
+    Sensitivity preconditioner class. 
+
+    In each call to the preconditioner the `gradient` is multiplied by :math:` 1/(A^T \mathbf{1})` where :math:`A` is an operator, :math:`\mathbf{1}` is an object in the range of the operator filled with ones.
+
+    Parameters
+    ----------
+    operator : CIL Operator 
+        The operator used for sensitivity computation.
+    reference : DataContainer e.g. ImageData, optional
+        The reference data, an object in the domain of the operator. Recommended to be a best guess reconstruction. 
```

Comment:
removed? Not sure why that happened. If you do WLS, this preconditioner would probably taking the weights into account, but now you cannot anymore

Reply:
Yep, I removed it from sensitivity. It is still included in AdaptiveSensitivity

Reply:
yes, but `AdaptiveSensitivity` also multiplies with the recon-image (I think), so it's a bit different. Naming problem as well I suppose.

Reply:
I changed adaptive sensitivity to have two options:
1. Sensitivity with reference  $ (x_{ref}+\delta)/(A^T1)$
2.  Adaptive $ (x_n+\delta)/(A^T1)$ which can then be freezed at a fixed number of iterations